### PR TITLE
fix(providers): Reduce memory allocation (gc thrash)

### PR DESCRIPTION
**Pull request by aaron-kasten**
This change causes Minio libraries to reduce memory allocations.  Specifically, buffer allocations for S3 upload.

### CODE REVIEW

**Comment by jkowalski** [comment]
this looks good but the code is very tricky and needs unit tests before we can merge

**Comment by aaron-kasten** [comment]
> this looks good but the code is very tricky and needs unit tests before we can merge

@jkowalski I'm unsure of how to unit test this.

**Comment by aaron-kasten** [comment]
image

This is an example of the memory allocation savings for a 5GB snapshot - approximately 3x the memory of the snapshot size in this case.

**Review by redgoat650 — COMMENTED** [review]
Looks good, just a suggestion for increasing test coverage:

I agree with @jkowalski's sentiment from his earlier review: "this code is very tricky". I manually inspected the logic and it _SEEMS_ fine, but it's very susceptible to off-by-one errors or regressions if not tested completely.

The unit test you added covers some corner+error cases, but seems to be missing some happy-path buffer filling and content-checking cases.

I've drawn up a diagram of some test cases that we might add for the `ReadAt` implementation. For each case, we could assert the expected number of bytes written, the correct `p []byte` resulting content (to ensure we're not starting from the wrong offset or leaving the output buffer incomplete), and whether it expects an error return.

The test cases are various permutations of buffer lengths and offsets aligning/overlapping in various ways, meant to catch any off-by-one errors that might get introduced.

I hope you find this sketch intelligible. I've included a "key" up top to explain what each proposed test case looks like. Each test case is comprised of a BLUE `WriteBuffer` which can consist of one or more `b.b.Slices`, a GREEN `p []bytes` target buffer, and an ORANGE `offset` value.

Feel free to incorporate any/all/none of the cases in the unit test; I'm just sharing some cases I would have considered adding, just to be extra sure of the implementation. Other than strengthening tests, as far as I can tell the change looks good!

image

**Comment by aaron-kasten** [comment]
@redgoat650 Lets chat later.  Test coverage has been enlarged to deal with corner cases not for Kopia ReaderAt

**Comment by aaron-kasten** [comment]
Im unclear on how the test cases you mention are specific to this implementation vs the Buffer implementation in general.  The thinking is that anything that applies to the original Buffer should be implemented (as its already been accepted).

**Comment by aaron-kasten** [comment]
Are you proposing that additional coverage be introduced for the original Buffer implementation?

**Comment by redgoat650** [comment]
Ah! I think I missed that all of the `ReaderAt` cases I mentioned are probably being tested by this line
```go
	require.NoError(t, iotest.TestReader(reader, buf))
```
Is that accurate, @aaron-kasten? If so, that's great. Perhaps we can add a comment to draw attention to it? It was easy to miss when reading through the test case.

**Comment by aaron-kasten** [comment]
A majority, at least for compliance with the Go apis. The scatter-gather buffer in Kopia is a different beast though.
Im wondering if it needs additional tests ... which your comments imply ... then it might be worth while expanding them.
Its later on a friday for me.  Ill take a closer look at the code and your comments on Monday and see what can be done.

**Comment by aaron-kasten** [comment]
Thank you for your detailed observations ... its a big help having something so specific to work from.

**Comment by aaron-kasten** [comment]
> Ah! I think I missed that all of the `ReaderAt` cases I mentioned are probably being tested by this line
> 
> ```go
> 	require.NoError(t, iotest.TestReader(reader, buf))
> ```
> 
> Is that accurate, @aaron-kasten? If so, that's great. Perhaps we can add a comment to draw attention to it? It was easy to miss when reading through the test, and I thought we were missing the obvious cases 😄
> 
> [edit to add the `TestReader` description]:
> 
> ```go
> // TestReader tests that reading from r returns the expected file content.
> // It does reads of different sizes, until EOF.
> // If r implements [io.ReaderAt] or [io.Seeker], TestReader also checks
> // that those operations behave as they should.
> //
> // If TestReader finds any misbehaviors, it returns an error reporting them.
> // The error text may span multiple lines.
> ```

@redgoat650 In the future can you submit this as a change - ideally inline as a comment with a change or a change directly to the branch?  It saves on risking my interpretation of your intended meaning and saves steps (Its implied to be approved by you and does not need to go through another review cycle).

File: internal/gather/gather_bytes_test.go
Comment (redgoat650) [inline internal/gather/gather_bytes_test.go:187]: As discussed with @Shrekster, we can probably make use of actual chunk size constants here. And we can iterate over a few key sizes, passing each into the iotest helper, just for an easy increase in coverage. I think @Shrekster will handle the quick enhancement, and I can approve.

**Review by jkowalski — APPROVED** [review]
LGTM, thanks for adding the test.

File: internal/gather/gather_bytes_test.go
Comment (redgoat650) [inline internal/gather/gather_bytes_test.go:292]: nit: since this is THE important line, behind which all of the tests will run, I'd make it stand out a bit more, instead of cuddling it with the reader extraction above. Maybe give it its own newline and comment?
```suggestion
			require.Truef(t, ok, "MUST implement io.ReaderAt, chunkSizeMultiplier: %f", chunkSizeMultiplier)
			
			// execute standard ReadAt tests.
			require.NoErrorf(t, iotest.TestReader(preRdr, buf),
```

File: internal/gather/gather_bytes_test.go
Comment (redgoat650) [inline internal/gather/gather_bytes_test.go:269]: So IIUC, we're testing contentBufs with size `[3.1x, 3.2x, ..., 3.9x]` (x is the chunk size)? I was thinking we could try more interesting contentBuf sizes like `[0.5x, 1x, 2x, 2x-1, 2x+1,...]` that sort of thing (x is the chunk size). That would probe around the edges of the different regimes involved in the underlying buffer implementation. The current test just exercises the "multiple slices with a partial slice at the end" case.

**Review by redgoat650 — APPROVED** [review]
Some suggestions for improvement, otherwise looks good

**Comment by Shrekster** [comment]
Added keep-open, label. I'd like to test this more thoroughly. Do not merge yet.

File: internal/gather/gather_bytes_test.go
Comment (Shrekster) [reply internal/gather/gather_bytes_test.go]: done

File: internal/gather/gather_bytes_test.go
Comment (Shrekster) [reply internal/gather/gather_bytes_test.go:297]: done

**Comment by Shrekster** [comment]
> Added keep-open, label. I'd like to test this more thoroughly. Do not merge yet.

done with testing, should be good now; fixed some slice index increment bugs

File: internal/gather/gather_bytes.go
Comment (Shrekster) [inline internal/gather/gather_bytes.go:170]: I should add to the comment here to make an important behaviour slightly more evident here:

“if we have not read the full curSlice then later on we will simply return n, nil and not try to work on the next slice”.

File: internal/gather/gather_bytes_test.go
Comment (redgoat650) [inline internal/gather/gather_bytes_test.go:264]: Comment can be updated I think
```suggestion
	// Test some interesting input buffer sizes from a 1-byte buffer to many multiples 
	// of the internal allocator chunk size.
```

File: internal/gather/gather_bytes_test.go
Comment (redgoat650) [inline internal/gather/gather_bytes_test.go:283]: we can remove the case if it's unused
```suggestion
```

**Review by redgoat650 — APPROVED** [review]
The added test cases look great, thanks @Shrekster ! And nice catch on the 1-chunk bug 🚀

**Review by julio-lopez — COMMENTED** [review]
@aaron-kasten Thanks for submitting the change, this is will result in significant memory usage reduction for users with repositories on S3 compatible storage. And kudos for profiling, diagnosing, tracking down the issue 🎖️ 🎩 and adding the tests.

And thanks to @redgoat650 and @Shrekster for the detailed review in this involved codepath, and for chipping in with the fixes.

Great job team.

Please feel free to merge once CI passes.

**Comment by julio-lopez** [comment]
I also triggered the storage provider tests. Please keep an eye on those before merging.

https://github.com/kopia/kopia/actions/runs/8504413695/job/23291280771?pr=3695

**Comment by aaron-kasten** [comment]
@julio-lopez (@Shrekster ) got an error in the code-coverage test: `EPIPE`.  Something happened when trying to upload coverage results (?).

Have either of you seen this before?  Is there a standard resolution?

```
gpg: directory '/home/runner/.gnupg' created
gpg: keybox '/home/runner/.gnupg/pubring.kbx' created
gpg: /home/runner/.gnupg/trustdb.gpg: trustdb created
gpg: key 806BB28AED779869: public key "Codecov Uploader (Codecov Uploader Verification Key) <security@codecov.io>" imported
gpg: Total number processed: 1
gpg:               imported: 1

Error: write EPIPE
    at afterWriteDispatched (node:internal/stream_base_commons:160:15)
    at writeGeneric (node:internal/stream_base_commons:151:3)
    at Socket._writeGeneric (node:net:952:11)
    at Socket._write (node:net:964:8)
    at writeOrBuffer (node:internal/streams/writable:447:12)
    at _write (node:internal/streams/writable:389:10)
    at Socket.Writable.end (node:internal/streams/writable:665:17)
    at Socket.end (node:net:722:31)
    at module.exports (/home/runner/work/_actions/codecov/codecov-action/54bcd8715eee62d40e33596ef5e8f0f48dbbccab/node_modules/gpg/lib/spawnGPG.js:50:1)
    at Object.call (/home/runner/work/_actions/codecov/codecov-action/54bcd8715eee62d40e33596ef5e8f0f48dbbccab/node_modules/gpg/lib/gpg.js:28:1)
```

**Comment by aaron-kasten** [comment]
@Shrekster @jkowalski @julio-lopez I made some final changes. Please take a look.

File: internal/gather/gather_bytes.go
Comment (julio-lopez) [inline internal/gather/gather_bytes.go]: @aaron-kasten IIUC, @Shrekster modified these 3 lines because he ran into a bug. Could you folks please double check?

I'm going to re-trigger the providers test in the mean time.

File: internal/gather/gather_bytes.go
Comment (Shrekster) [inline internal/gather/gather_bytes.go:178]: what happens if `maxBsIndex = n` ? We probably do not want to enter this loop then, right ?

Comment (Shrekster) [reply internal/gather/gather_bytes.go:178]: I think the cop() below is probably saving you from a panic here, but better to not even enter this loop when we have filled the input buffer

File: internal/gather/gather_bytes.go
Comment (redgoat650) [inline internal/gather/gather_bytes.go:182]: perhaps we can make the code more self-describing in addition to comments?
```suggestion
	for moreRoomInOutput(n, maxBsIndex) && moreSlicesLeft(sliceNdx, slicesN) {
```

**Review by redgoat650 — APPROVED** [review]
Thanks for adding clarity. One optional suggestion, but otherwise looks good

**Comment by julio-lopez** [comment]
> @julio-lopez (@Shrekster ) got an error in the code-coverage test: `EPIPE`. Something happened when trying to upload coverage results (?).

Likely an infra issue on their side?? dunno. 🤷🏼

**Comment by jkowalski** [comment]
🎉

**Pull request merged**
